packages/inferno-router/src/Router.ts

```diff
@@ -1,46 +1,89 @@
 import { Component, Inferno } from 'inferno';
+import type { History, Location } from 'history';
+import { combineFrom, isUndefined } from 'inferno-shared';
 import { warning } from './utils';
-import { combineFrom } from 'inferno-shared';
-import type { History } from 'history';
+import { Match } from './Route';
+import { resolveLoaders, traverseLoaders } from './resolveLoaders';
+
+export type TLoaderProps<P extends Record<string, string>> = {
+  params?: P; // Match params (if any)
+  request: Request; // Fetch API Request
+  // https://github.com/remix-run/react-router/blob/4f3ad7b96e6e0228cc952cd7eafe2c265c7393c7/packages/router/router.ts#L1004
+  // https://github.com/remix-run/react-router/blob/11156ac7f3d7c1c557c67cc449ecbf9bd5c6a4ca/examples/ssr-data-router/src/entry.server.tsx#L66
+  // https://github.com/remix-run/react-router/blob/59b319feaa12745a434afdef5cadfcabd01206f9/examples/search-params/src/App.tsx#L43
+  // https://github.com/remix-run/react-router/blob/11156ac7f3d7c1c557c67cc449ecbf9bd5c6a4ca/packages/react-router-dom/__tests__/data-static-router-test.tsx#L81
+  
+  onProgress?(perc: number): void;
+}
+
+/**
+ * Loader returns Response and throws error
+ */
+export type TLoader<P extends Record<string, string>, R extends Response> = ({params, request}: TLoaderProps<P>) => Promise<R>;
+
+export type TLoaderData = {
+  res?: Response;
+  err?: any;
```

Comment:
Error type?

Reply:
This is unknown at this point since the actual implementation of the error is up to the developer (could be anything from a simple error message string, a JSON-object or an actual error object).

Ref: https://github.com/infernojs/inferno/pull/1621#discussion_r1060147378